UKHomeOffice/nmsw-ui
# NMSW-30 Add validation, error handling, and error summary

# Ticket

NMSW-30

----

## AC

GIVEN I am viewing the sign-in form
WHEN I attempt to submit the form without providing an email address or password
THEN  I am shown form level errors

GIVEN I have attempted to submit the sign in form
WHEN I have not provided an email address that has an '@' symbol in it
THEN I am shown a form level error related to you must provide a valid email address

GIVEN I have attempted to submit the sign in form
WHEN I have not provided a password that meet minimum requirements
THEN I am shown a form level error related to an invalid password

GIVEN I am viewing the error panel
WHEN I click an error link in that panel
THEN I am taken to the field with the error

----

## To test

- run app locally
- go to localhost:3000
- click start now
- > you should see the sign in form
- click submit without adding anything to the fields
- > you should see the error banner with 'required' errors for all fields
- enter text in the email address field without the @ symbol
- do not enter text in the password field
- click submit
- > you should see an error that the email address is invalid
- > you should see an error that the password is required
- > you should see an error that the sample field is required
- click on the email address error
- > you should be scrolled to the email field question/label
- the email field should be focused
- enter a valid format email address
- enter something in the password field
- enter 2 characters in the 'sample min length' field
- click submit
- > you should see an error that the field is too short
- enter a valid string
- click submit
- > you should see no errors
- > you should be taken to the dashboard

----

## Notes

NOTE: As sign in does not give instructions on password length/structure I have added an extra sample field for testing purposes to test having a required & min length field. We can remove this once we finish with foundations and start building true user facing forms.
